crates/driver/src/domain/competition/quote.rs

```diff
@@ -0,0 +1,182 @@
+                buy: fulfillment.order.buy,
+                interactions,
+            },
+        }
+    }
+}
+
+/// An order which needs to be quoted.
+#[derive(Debug)]
+pub struct Order {
+    pub sell_token: eth::TokenAddress,
+    pub buy_token: eth::TokenAddress,
+    pub amount: Amount,
+    pub valid_to: u32,
+    /// See [`crate::domain::competition::order::Partial`].
+    pub partial: bool,
+    pub quality: Quality,
+    pub gas_price: eth::EffectiveGasPrice,
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum Amount {
+    Sell(eth::U256),
+    Buy(eth::U256),
+}
+
+impl Order {
+    /// The asset being bought. Zero if this is a sell.
+    pub fn buy(&self) -> eth::Asset {
+        match self.amount {
+            Amount::Sell(..) => eth::Asset {
+                amount: Default::default(),
+                token: self.buy_token,
+            },
+            Amount::Buy(amount) => eth::Asset {
+                amount,
+                token: self.buy_token,
+            },
+        }
+    }
+
+    /// The asset being sold. Zero if this is a buy.
+    pub fn sell(&self) -> eth::Asset {
+        match self.amount {
+            Amount::Sell(amount) => eth::Asset {
+                amount,
+                token: self.sell_token,
+            },
+            Amount::Buy(..) => eth::Asset {
+                amount: Default::default(),
+                token: self.sell_token,
+            },
+        }
+    }
+
+    pub fn side(&self) -> competition::order::Side {
+        match self.amount {
+            Amount::Sell(..) => competition::order::Side::Sell,
+            Amount::Buy(..) => competition::order::Side::Buy,
+        }
+    }
+}
+
+/// Quality of the quote to be generated. This value determines the time
+/// allocated for the solver to generate the solution.
+#[derive(Debug, Clone, Copy)]
+pub enum Quality {
+    Fast,
+    Optimal,
+}
+
+impl Order {
+    /// Generate a quote for this order. This calls `/solve` on the solver with
+    /// a "fake" auction which contains a single order, and then determines
+    /// the quote for the order based on the solution that the solver
+    /// returns.
+    pub async fn quote(&self, solver: &Solver, config: &Config) -> Result<Quote, Error> {
+        let solution = solver
+            .solve(&self.fake_auction(), self.timeout(config))
+            .await?;
+        let fulfillment = solution
+            .trades
+            .into_iter()
+            .filter_map(|trade| match trade {
+                competition::solution::Trade::Fulfillment(fulfillment) => Some(fulfillment),
+                competition::solution::Trade::Jit(..) => None,
+            })
+            .exactly_one()
+            .map_err(|_| Error::QuotingFailed)?;
+        Ok(Quote::new(fulfillment, solution.interactions))
+    }
+
+    fn fake_auction(&self) -> competition::Auction {
+        competition::Auction {
+            id: None,
+            tokens: Default::default(),
```

Comment:
I believe this might build the wrong fake auction. This build an auction that the driver expects in the `/solve` request; not the kind of auction that gets sent to the `solver`, right?
If it's easier one could consider building the other kind of auction. This is also what the current http estimator does.

Reply:
I theory there could be a difference. Specifically:
- It is fine to `/quote` with "recent" liquidity (i.e. liquidity whose state was updated recently, but not necessarily on the "latest" known block)
- Solving should be done with "latest" liquidity.

Note that this is a performance optimization and doesn't have to be done for the solver+driver colocation release, just something to keep in mind that we may need separate logic in the driver for quoting eventually.

Reply:
I also think it can come later:
> Note that this is a performance optimization and doesn't have to be done for the [initial] solver+driver colocation release